etc/docs/Design.md

```diff
@@ -0,0 +1,100 @@
+
+* ExecuteCommand: Starts a command with limited resources in its own namespace with the supplied arguments
+* StopExecution : Stop a certain execution 
+* ListExecutions: List all the executions the connected client has rights to
+* GetExecution: Gets metadata about a certain execution defined by an execution ID
+* GetOutStream: Get a stream of lines of standard output (stdout & stderr) of a certain execution
+* GetOut: Get all lines of standard output (stdout & stderr) that has been written so far by the process
+
+To read about the exact details of the API you can look at [the proto file](api.proto)
+
+### Streaming data
+
+In order to be able to stream and also send the full output of the processes we need to store the output of this in memory for simplicty, in order to be able to keep memory from being overloaded a simple recylcing of these processes will be implemented. See [Tradeoffs](#tradeoffs) section for information about why this solution should not be used in production  
+
+## Security
+
+### Cipher suite
+
+In order to be secure the app will use the TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 cypher suite.
+
+It suits our needs to be a strong cypher suite by using AES, SHA-2 and ephemeral Diffie-Hellman key exchange, even though the ephemeral key exchange thakes more compute power, with the amount of the frequency of the connections (processes are started, then waited on, and output read) the added security is worth it
+
+### Authentication
+
+### mTLS
+
+For the mTLS the following certificates will be generated by a generate-certs shell script to enable easy testing of the application (see the Tradeoff section for more details):
+
+* A root CA certificate to sign client and server CSR-s 
+* A x509 certificate for the server signed by the root CA
+* Two x509 certificates one for each of two clients, that could be used to connect to the applicaiton
+
+The root CA will use a private key that's 4096 bits and the client and server certs will use private keys with 2048 bits
+
+### Authorization
+
+To keep things simple, the following statement should hold for every process started on the server:
+
+`Only clients authenticated with the client certs hacing the same CN can get any information about a certain process`
+
+## Process execution
+
+### Resource control
+
+To control resource allocation linux control groups (chroups) can be utilized
+
+For the required CPU, Memory, and DiskIO limits a new cgroup could be created per process using the cpu, memory and blkio controllers
+
+### Isolation
+
+To implement isolation of namespaces for PID, Mount and Networks unix namespaces will be used
+
+In order to be able to set up proper isolation there is need to run some initialization process before the app executes in the new namespace.
+
+## Client
+
+The client CLI will have a two top level commands:
+
+`start` and `execution`
+
+The `start` command will send a command to be started to the server, along with the arguments passed to it
```

Comment:
Could you give a few CLI examples of how these commands will be used? Think like a CLI --help output to show commands/args/flags.